java-checks/src/main/java/org/sonar/java/checks/InappropriateRegexpCheck.java

```diff
@@ -0,0 +1,83 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodMatcher;
+import org.sonar.java.model.LiteralUtils;
+import org.sonar.java.tag.Tag;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2639",
+  name = "Inappropriate regular expressions should not be used",
+  priority = Priority.CRITICAL,
+  tags = {Tag.BUG})
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.INSTRUCTION_RELIABILITY)
+@SqaleConstantRemediation("15min")
+@ActivatedByDefault
+public class InappropriateRegexpCheck extends AbstractMethodDetection {
+
+  private static final String INAPPROPRIATE_REGEXPS = "\\.|\\|";
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodMatcher.create().typeDefinition("java.lang.String").name("replaceAll").withNoParameterConstraint()
+      );
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree firstArg = mit.arguments().get(0);
+    if (isInappropriateRegexpStringLiteral(firstArg) || isFileSeparator(firstArg)) {
+      reportIssue(firstArg, "");
+    }
+  }
+
+  private static boolean isInappropriateRegexpStringLiteral(ExpressionTree firstArg) {
+    if (firstArg.is(Tree.Kind.STRING_LITERAL)) {
+      String regexp = LiteralUtils.trimQuotes(((LiteralTree) firstArg).value());
+      return regexp.matches(INAPPROPRIATE_REGEXPS);
+    }
+    return false;
+  }
+
+  private static boolean isFileSeparator(ExpressionTree firstArg) {
+    if (firstArg.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) firstArg;
+      return mse.identifier().name().equals("separator") && mse.expression().symbolType().is("java.io.File");
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Move the "separator" string literal on the left side of this string comparison. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1132)